bitwarden_rs: 1.9.1 -> 1.13.0

Br1ght0ne opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
https://github.com/dani-garcia/bitwarden_rs/releases/tag/1.13.0
MySQL and (initial) PostgreSQL support have been added, and are accomodated in this PR. The default DB is still SQLite (to preserve backwards compatibility).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
```
/nix/store/rvmzjirrbdmy9n042ssk72qkami10yij-bitwarden_rs-1.9.1	  47.2M
```
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @msteen

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/nixos/nixpkgs/74927)
<!-- Reviewable:end -->

Br1ght0ne committed:
bitwarden_rs: 1.9.1 -> 1.13.0

Co-authored-by: msteen <msteen@users.noreply.github.com>
Co-authored-by: Daniel Løvbrøtte Olsen <daniel.olsen99+GitHub@gmail.com>

Br1ght0ne committed:
bitwarden_rs: add filalex77 to maintainers

msteen commented:
Thank you for updating bitwarden_rs! I have not yet tested the changes myself, but I have looked over the code changes and they look good to me. However what are the implications on the [corresponding bitwarden_rs module](https://github.com/NixOS/nixpkgs/tree/master/nixos/modules/services/security/bitwarden_rs)? Doesn't the backup script and its service `PATH` need to change depending on the database backend used?

Ma27 reviewed (changes requested):
Building `bitwarden_rs` breaks for me with the following failure:

```
bitwarden_rs>    Compiling bitwarden_rs v1.0.0 (/build/source)
bitwarden_rs>    Compiling subtle v2.2.2
bitwarden_rs> error: You need to enable one DB backend. To build with previous defaults do: cargo build --features sqlite
bitwarden_rs>   --> build.rs:12:5
bitwarden_rs>    |
bitwarden_rs> 12 |     compile_error!("You need to enable one DB backend. To build with previous defaults do: cargo build --features sqlite");
bitwarden_rs>    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
bitwarden_rs>    Compiling data-encoding v2.1.2
bitwarden_rs>    Compiling quoted_printable v0.4.1
bitwarden_rs> error: aborting due to previous error
bitwarden_rs> error: could not compile `bitwarden_rs`.
bitwarden_rs> warning: build failed, waiting for other jobs to finish...
bitwarden_rs> error: build failed
bitwarden_rs> builder for '/nix/store/znfdg3299j5mhq87y1p6mdxlc92l41vg-bitwarden_rs-1.13.0.drv' failed with exit code 101
```

Br1ght0ne committed:
nixos/bitwarden_rs: add support for db backends

Rolling out initial support for:
- mysql (1.10.0)
  https://github.com/dani-garcia/bitwarden_rs/releases/tag/1.10.0
- postgresql (1.11.0)
  https://github.com/dani-garcia/bitwarden_rs/releases/tag/1.11.0

As current backup strategy is sqlite-specific, more work is needed.
For now, an assertion to make users know those wont work.

filalex77 commented:
@msteen I updated the module (didn't implement alternative backup strategies yet).

Ma27 commented on nixos/modules/services/security/bitwarden_rs/default.nix:
```diff
@@ -23,10 +23,21 @@ let
     enable = mkEnableOption "bitwarden_rs";
 
+    dbBackend = mkOption {
+      type = enum [ "sqlite" "mysql" "postgresql" ];
```
I personally think that this is pretty misleading. When I see allowed values e.g. in a man-page, I'd expect all of them to work. So we should either update the description accordingly or (preferably) leave out `mysql` and `postgresql` for now.

msteen replied:
I agree, it would be best to simply the PR and just force the use of the sqlite backend for now. That way version 1.13 can already be used, and once people truly want to start different backends, proper attention can be put into it to make sure they are actually completely supported, like in the backup script and maybe other parts.

filalex77 replied:
@Ma27 @msteen Is the current implementation all right?

msteen replied:
I shall test the the changes once I am home, but I would prefer the files to be changed to not include the other databases and not make sqlite optional, but required, since for now we make it the required backend until further testing / requests are made.

msteen replied:
There seems to be a bug in the rust package build script because it adds the test flags after `--`, causing them not to be interpreted as flags, so the `--features` argument that it passed to it ends up being ignored. I will make a PR or a bug report for this later. For now it can be disabled by forcibly setting `checkPhase` to the empty string, because `dontCheck` does not seem to exist for rust packages, it has `doCheck` instead, but it seems to be ignored for the `checkPhase`, which is run regardless of what `doCheck` is set too. [Here is a gist](https://gist.github.com/msteen/e126394d9b70d553ca7aca0f3c8eb113) with how I would change the package for now, i.e. no need to change the module file, remove the patch, and update package definition as in the gist, which is the same as you had it without the bit about the databases, since we are defaulting to `sqlite` for now anyway, which does not need any additional packages. I have tested it on my desktop and it seems to work fine as far as I have tested it.

Ma27 commented:
Ah sorry, I totally forgot about that one! (next time feel free to ping me!).

Will review/test it locally now :)

msteen reviewed (changes requested):
Since we are hardcoding the use of sqlite for now, it is better not to commit the work in progress done in the module, none of those changes are necessary at the moment.

Ma27 reviewed (commented):
The module currently breaks with the following error:

```
error: anonymous function at /home/ma27/Projects/nixpkgs/pkgs/tools/security/bitwarden_rs/default.nix:1:1 called with unexpected argument 'dbBackend', at /home/ma27/Projects/nixpkgs/lib/customisation.nix:69:16
(use '--show-trace' to show detailed location information)
```

msteen commented on pkgs/tools/security/bitwarden_rs/default.nix:
```diff
@@ -2,32 +2,31 @@
+  cargoBuildFlags = [ "--features sqlite" ];
+
+  # FIXME: The rust check phase seems broken due to passing the flags after --.
+  doCheck = false;
```
Well, the right fix would be to fix `checkPhase` instead, because it does not really make sense to add flags after `--`. Apparently some flags can be added after `--` (seems a weird API design choice to me) as can be seen in the `cargo test` manpage (e.g. `cargo test -- --nocapture`), but the documented flags have to be added before `--`, like `--features`. Maybe it can be fixed by adding `checkArgs` to differentiate between flags and arguments.

Mic92 replied:
Just passing all cargo flags to cargo test could solve this issue.

Mic92 commented:
This is what I did to implement postgresql/mysql as well: https://github.com/Mic92/nixpkgs/commit/86031cd438d096c497ff59c72202a6a7df919214

I had to switch to rust nightly in my fork so.

Ma27 commented:
> I had to switch to rust nightly in my fork so.

Do we actually need nightly features or is it sufficient to set `RUSTC_BOOTSTRAP`?

Mic92 commented:
> > I had to switch to rust nightly in my fork so.
> 
> Do we actually need nightly features or is it sufficient to set `RUSTC_BOOTSTRAP`?

Ah. This might be the problem.

Mic92 commented:
This variant no longer needs nightly: 30c49ee7ba3c43b13f1e26ae8dc7ef6b03e3d13a

filalex77 commented:
Closing in favor of #78615.
